cou/apps/auxiliary.py

```diff
@@ -457,3 +462,146 @@ async def _verify_nova_compute(self, target: OpenStackRelease) -> None:
+        cacert_file = None
+        if cacert_b64:
+            with tempfile.NamedTemporaryFile(mode="wb", delete=False) as fp:
+                fp.write(base64.b64decode(cacert_b64))
+                cacert_file = fp.name
+        return cacert_file
+
+    async def _get_unit_api_url(self, unit_name: str) -> str:
+        """Get unit's vault api address.
+
+        :param unit_name: vault unit name
+        :type unit_name: str
+        :return: unit's vault api url
+        :rtype: str
+        """
+        # Use vip as address if vip is being used.
+        if self.config["vip"].get("value"):
+            address = self.config["vip"].get("value")
+        else:
+            juju_unit = await self.model.get_unit(unit_name)
+            address = juju_unit.public_address
+
+        transport = "http"
+        if self.config["ssl-cert"].get("value"):
+            transport = "https"
+        return f"{transport}://{address}:8200"
+
+    async def _wait_for_sealed_status(self) -> None:
+        """Wait for application vault go into sealed status.
+
+        :raises ApplicationError: When application vault is not in sealed status.
+        """
+        await self.model.wait_for_idle(
+            timeout=self.wait_timeout,
+            status="blocked",
+            apps=[self.name],
+            # Vault application will go into error status first then blocked status.
+            raise_on_error=False,
+        )
+
+        app_status = await self.model.get_application_status(charm_name="vault")
+        if not app_status.status.info == "Unit is sealed":
+            raise ApplicationError("Vault not in sealed status")
+        logger.debug("Application 'vault' in sealed status")
+
+    async def _get_vault_client(self, unit_name: str) -> hvac.Client:
+        """Get vault client.
+
+        :param unit_name: vault unit name
+        :type unit_name: str
+        :return: hvac vault Client object
+        :rtype: hvac.Client
+        """
+        vault_url = await self._get_unit_api_url(unit_name)
+        cacert_file = self._get_cacert_file()
+
+        client = hvac.Client(url=vault_url, verify=cacert_file)
+        return client
+
+    async def _unseal_vault(self) -> None:
+        """Unseal vault on every vault unit.
+
+        :raises VaultGetStatusFailed: When failed to get sealed status from vault command
+        :raises VaultUnsealFailed: When unseal command failed
+        """
+        progress_indicator.stop()
+
+        for unit_name in self.units:
+            logger.debug("Start unseal %s", unit_name)
+            client = await self._get_vault_client(unit_name)
+            while True:
+                status = client.sys.read_seal_status()  # pylint: disable=no-member
+                if status["sealed"]:
+                    unseal_key = getpass.getpass("Unseal Key (will be hidden):")
+                    if unseal_key:
+                        client.sys.submit_unseal_key(key=unseal_key)  # pylint: disable=no-member
```

Comment:
I guess it's necessary to check the json response and see if it was 200. If user pass a wrong key, As far as I understood it won't ask again if the key passed is wrong

Reply:
Not necessary, because even `submit unseal key` failed. The vault will still be in sealed status, and user need to input again.